src/mvs/fusion.cc

```diff
@@ -383,25 +440,35 @@ void StereoFusion::Fuse() {
     // Set the current pixel as visited.
     fused_pixel_mask.Set(row, col, true);
```

Comment:
You will end up with race conditions here, because you simultaneously read/write to the masks from multiple threads.

Reply:
Yes this is true, and ideally this should have been an atomic update, but unfortunately the OpenMP standard (at least on Windows) does not allow atomic operations with function calls, only simple updates like `x = val` or `x += val`, etc. Wrapping this in a critical section kills the performance since all threads get blocked instead of just the ones that may conflict so this was also a no-go from a performance standpoint.

I've let that race condition exist since all threads are only attempting to set the same value (i.e. make the mask value `true`) so they're not adversely impacting the result of fusion. Worst case, by my understanding, is that we reconsider a pixel that should have already been masked out as visited. This causes some potential duplication of work, but from a performance standpoint its practically negligible, and definitely much preferred to using a critical section.

If there is a better way to go about it I can update accordingly. If it so happens that atomic works better in Linux/Mac I can add the `#pragma omp atomic` with a specific guard to skip it on windows. Otherwise I would have to change the data type of the mask from `Mat<bool>` to `std::vector` or something similar so I can directly access the data without a need for a function call.

Reply:
I think I managed to make this work with `#pragma omp atomic` by accessing the underlying pointer of `Mat<bool>`. Of course I'll have to performance test it on my end to ensure there is no adverse performance impact.

Reply:
What I tried won't work either. Apparently `std::vector<bool>` does not have a `.data()` member, unlike other types of vectors. I managed to make it work by converting the mask to `Mat<int>` instead of `Mat<bool>` and doing this:
```
const int pixel_index = depth_map_sizes_.at(image_idx).first * row + col;
#pragma omp atomic
fused_pixel_mask.GetPtr()[pixel_index]++;
```

Note that I'm using increment and not assignment here since Windows only support OpenMP 2.0 and atomic only supports `+=`, `-=`, `++`, and `--`.

If you think that's a valid way of handling this then I'll go ahead and update the PR. Thankfully making this atomic does not impact performance for parallel execution, at least in the few datasets I tested.

Reply:
I think your suggested fix doesn't resolve the actual problem causing potentially many duplicated fused points. I think this effect is exacerbated by the fact that you start spawning threads for each pixel row, i.e. nearby pixels will be processed by different threads, which are very likely to be fused into the same point.

I think we could minimize this effect by using some striding between different threads. For example:
```
const int kRowStride = max(1, height / num_threads);
#pragma omp parallel for schedule(dynamic) num_threads(num_threads)
for (int row_start = 0; row_start < height; row_start += kRowStride) {
    const int row_end = min(height, row_start + kRowStride);
    for (int row = row_start; row < row_end; ++row) {
        for (int col = 0; col < width; ++col) {
            ...
        }
    }
}
```
In this case, elements in the fusion queue of different threads are further apart and are less likely to race with each other.

I was not so much concerned about the atomic update, which will obviously be a problem for `bool`, which is typically implemented using bitshifting. We could switch to `Mat<char>` and just use `mask.Set(row, col, 1)`.

Reply:
Yes, you are right, the current setup with dynamic scheduling does suffer from potential duplications. I think an easier approach could be to configure it with a batch size of 10 or more so that different threads start further apart; this avoids the need for manual management of offsets. I can experiment with a few different batch sizes and depth map sizes and see if there's some value that works well enough.

And yes doing `Mat<char>` for mask should work even without a specific guard for the atomic update, which will keep things simpler.

These two I can do and quickly update, but changing the format of fusion to work with the existing thread pool will take a while as it's not clear to me what kind of setup it will need to process the problems, and I don't want to make the approach too verbose. But, I'll definitely attempt this at some point.

Reply:
So I just tested the current setup (chunk size of 1) against the single threaded run and multi-threaded runs with chunk sizes of 10, 20, and 50 (that should be the same as setting a row stride).

Interestingly, the processing time goes slightly up when increasing the chunk size, and the reconstructed points are slightly less than the single threaded point clouds, with the exception of chunk size of 1. Below is a summary of the findings.

- Reference single threaded run: Runtime 11.5 min; points: 7,570,936

The comparison results show the relative number of points and the % change:
1. Chunk size 1 (current schedule): Runtime: 1.45 min; points: +20,568 (+0.27%)
2. Chunk size 10: Runtime: 1.50 min; points: -13353 (-0.18%)
3. Chunk size 20: Runtime: 1.66 min; points: -6130 (-0.08%)
4. Chunk size 50: Runtime: 2.09 min; points: -2552 (-0.03%)

Those differences are really small, thankfully, and in light of that I think it would make sense to configure the scheduling with a chunk size of 10.

Reply:
Chunk size of 10 seems like a good trade off. Thanks!

Reply:
No need to do the thread pool - was just a suggestion.

Reply:
Seeing how colmap does not natively use OpenMP in favor of the ThreadPool, I'm afraid that a solution with OpenMP would pollute the code with quite a few guards checking if OpenMP is available.

I think I have a clear idea now on how to go about this with the ThreadPool, so I'll give it a try and update the PR. Should be cleaner overall with that approach.

Reply:
Looks great, thanks.